### PR TITLE
feat: avatar movement using double click

### PR DESCRIPTION
**Pull request by olavra**
# Pull Request Description
Now you can click anywhere in the world and the avatar will try to reach the destination in a straigh line. A marker determines the target location. You can jump, double jump, glide and shift to run while automatically reaching the destination.
The avatar will stop after a few milliseconds if there is no sustancial increment in movement in the desired direction (cant reach it).

This does not use Navmesh Agents so its pretty simple straight movement, nothing fancy.

## Test Instructions
- Double click anywhere, a marker will appear
- The avatar will try to move in that direction
- The marker should dissapear after reaching the destination or failing to reach.
- Only one marker visibile at a time
- You can shift to run , space to jump and glide while automatically walking.
- Open Controls menu to see new input double click on mouse left

### CODE REVIEW

**Comment by AnsisMalins** [comment]
## [feat: avatar movement using double click](https://github.com/decentraland/performance-testing/actions/runs/23873146434)
PR number: [7978](https://github.com/decentraland/unity-explorer/pull/7978)
[Windows change build](https://explorer-artifacts.decentraland.org/@dcl/unity-explorer/branch/feat/mouse-move/pr-20581-7bf9ae7/Decentraland_windows64.zip)
[Windows baseline build](https://explorer-artifacts.decentraland.org/@dcl/unity-explorer/releases/v0.137.0-alpha/Decentraland_windows64.zip)
[macOS change build](https://explorer-artifacts.decentraland.org/@dcl/unity-explorer/branch/feat/mouse-move/pr-20581-7bf9ae7/Decentraland_macos.zip)
[macOS baseline build](https://explorer-artifacts.decentraland.org/@dcl/unity-explorer/releases/v0.137.0-alpha/Decentraland_macos.zip)

### Intel Core i3

Change samples: 2671
Baseline samples: 2675

| Metric | Change | Baseline | Delta | Improvement | |
| --- | ---: | ---: | ---: | ---: | :---: |
| CPU average | 33.6 ms | 33.6 ms | 0.0 ms | -0.1% | ⚪ |
| CPU 1% worst | 55.4 ms | 54.3 ms | 1.0 ms | -1.9% | ⚪ |
| CPU 0.1% worst | 118.3 ms | 120.4 ms | -2.1 ms | 1.8% | ⚪ |
| GPU average | 15.4 ms | 14.7 ms | 0.7 ms | -4.5% | 🔴 |
| GPU 1% worst | 21.5 ms | 21.1 ms | 0.4 ms | -1.7% | ⚪ |
| GPU 0.1% worst | 22.6 ms | 23.0 ms | -0.4 ms | 1.7% | ⚪ |

File: Explorer/Assets/DCL/AvatarRendering/AvatarShape/Assets/DestinationMarkerView.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    /// Also cycles the emission colour through a full hue rainbow.
+    /// The AnimatorController is expected to be pre-assigned on the prefab's Animator.
+    /// </summary>
+    public class DestinationMarkerView : MonoBehaviour
+    {
+        private static readonly int START_STATE = Animator.StringToHash("Start");
+        private static readonly int IDLE_STATE = Animator.StringToHash("Idle");
+        private static readonly int EMISSION_COLOR = Shader.PropertyToID("_EmissionColor");
+
+        [SerializeField] public float rainbowSpeed = 1f;
```
Comment (daniele-dcl) [inline Explorer/Assets/DCL/AvatarRendering/AvatarShape/Assets/DestinationMarkerView.cs]: Usually we go with a property with private setter and use `[field: SerializeField]` on it

File: Explorer/Assets/DCL/AvatarRendering/AvatarShape/Assets/DestinationMarkerView.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        {
+            UpdateAnimation();
+            UpdateRainbow();
+        }
+
+        private void UpdateAnimation()
+        {
+            if (transitionedToIdle || animator == null) return;
+
+            AnimatorStateInfo info = animator.GetCurrentAnimatorStateInfo(0);
```
Comment (daniele-dcl) [inline Explorer/Assets/DCL/AvatarRendering/AvatarShape/Assets/DestinationMarkerView.cs]: I am not 100% sure if I understood this correctly. It seems the goal is to transition the animator to a given state after a while?

I believe we would want to do that with the exit time of the default state of the animator instead?

Is there a good reason to do it programmatically?

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+    {
+        private SingleInstanceEntity camera;
+        private ICharacterControllerSettings cachedSettings;
+        private InputAction sprintAction;
+        private InputAction walkAction;
+
+        private readonly GameObject markerPrefab;
+
+        // Stores the time of the previous left-click to detect double-clicks.
+        private float lastClickTime = float.MinValue;
```
Comment (daniele-dcl) [inline Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs]: It would be nice to check if it's possible to do this in the input actions asset by using a multi-tap interaction.

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+                ? Object.Instantiate(markerPrefab)
+                : CreateFallbackMarker();
+
+            targetMarker.name = "PointAndClickMarker";
+            targetMarkerView = targetMarker.GetComponent<DestinationMarkerView>() ?? targetMarker.AddComponent<DestinationMarkerView>();
+            targetMarker.SetActive(false);
+        }
+
+        protected override void Update(float t)
+        {
```
Comment (daniele-dcl) [inline Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs:91]: I recommend breaking down the update into private helpers for readability. It's clearly handling multiple tasks.

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+            targetMarkerView.ResetMarker();
+        }
+
+        private void HideMarker()
+        {
+            targetMarker.SetActive(false);
+        }
+
+        private static GameObject CreateFallbackMarker()
+        {
```
Comment (daniele-dcl) [inline Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs:169]: We could log an error if we have to fall back to this marker?

Seems to me it could only happen because of a config error.

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+            var component = new PointAndClickMovementComponent
+            {
+                TargetPosition = clickTarget,
+                StuckCheckElapsed = 0f,
+                PositionAtLastStuckCheck = World.Get<CharacterTransform>(entity).Position,
+            };
+
+            if (World.Has<PointAndClickMovementComponent>(entity))
+                World.Set(entity, component);
```
Comment (daniele-dcl) [inline Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs]: IIRC we have `AddOrSet`.

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+                && mouse.leftButton.wasPressedThisFrame)
+            {
+                float now = UnityEngine.Time.unscaledTime;
+
+                if (now - lastClickTime <= cachedSettings.PointAndClickDoubleClickThreshold)
+                {
+                    Vector2 screenPos = mouse.position.ReadValue();
+                    Ray ray = cameraComponent.Camera.ScreenPointToRay(new Vector3(screenPos.x, screenPos.y, 0f));
+
+                    if (Physics.Raycast(ray, out RaycastHit hit))
```
Comment (daniele-dcl) [inline Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs]: Maybe we can consider filtering this out by layer? Maybe also max distance to avoid super far away clicks?

File: Explorer/Assets/DCL/Landscape/Assets/GPUI/Grass/StylizedGrass_Indirect.mat [hunk 8]
```diff
@@ -14,8 +14,12 @@ Material:
   m_ValidKeywords:
   - _ADVANCED_LIGHTING
   - _FADING
+  - _GPU_GRASS_BATCHING
```
Comment (daniele-dcl) [inline Explorer/Assets/DCL/Landscape/Assets/GPUI/Grass/StylizedGrass_Indirect.mat]: Are changes to this material intentional?

**Review by daniele-dcl — DISMISSED** [review]
Besides the 1 change to the grass material that seems unintentional, everything else is minor 👍

**Comment by daniele-dcl** [comment]
@claude review

**Comment by AnsisMalins** [comment]
PR [#7978](https://github.com/decentraland/unity-explorer/pull/7978), run [#23909705745](https://github.com/decentraland/performance-testing/actions/runs/23909705745)

Builds: [Windows change](https://explorer-artifacts.decentraland.org/@dcl/unity-explorer/branch/feat/mouse-move/pr-20608-5c98770/Decentraland_windows64.zip), [Windows baseline](https://explorer-artifacts.decentraland.org/@dcl/unity-explorer/releases/v0.137.0-alpha/Decentraland_windows64.zip), [macOS change](https://explorer-artifacts.decentraland.org/@dcl/unity-explorer/branch/feat/mouse-move/pr-20608-5c98770/Decentraland_macos.zip), [macOS baseline](https://explorer-artifacts.decentraland.org/@dcl/unity-explorer/releases/v0.137.0-alpha/Decentraland_macos.zip)

### Intel Core i3

| Metric | Change | Baseline | Delta | Improvement | |
| --- | ---: | ---: | ---: | ---: | :---: |
| Samples | 2652 | 2673 | | | |
| CPU average | 33.9 ms | 33.6 ms | 0.3 ms | -0.8% | ⚪ |
| CPU 1% worst | 69.5 ms | 54.6 ms | 14.9 ms | -27.2% | 🔴 |
| CPU 0.1% worst | 254.0 ms | 142.5 ms | 111.4 ms | -78.2% | 🔴 |
| GPU average | 15.7 ms | 16.1 ms | -0.4 ms | 2.3% | ⚪ |
| GPU 1% worst | 23.2 ms | 21.1 ms | 2.2 ms | -10.3% | 🔴 |
| GPU 0.1% worst | 29.0 ms | 24.0 ms | 4.9 ms | -20.5% | 🔴 |

File: Explorer/Assets/DCL/AvatarRendering/AvatarShape/Assets/DestinationMarkerView.cs (hunk 1 above)
Comment (olavra) [reply Explorer/Assets/DCL/AvatarRendering/AvatarShape/Assets/DestinationMarkerView.cs]: Done in the last comit

File: Explorer/Assets/DCL/AvatarRendering/AvatarShape/Assets/DestinationMarkerView.cs (hunk 2 above)
Comment (olavra) [reply Explorer/Assets/DCL/AvatarRendering/AvatarShape/Assets/DestinationMarkerView.cs]: I was having problems with the animator controller transition and end up doing it programatically. It was a problem with the animation, so reverted this and made the transition directly using the exit time.

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs (hunk 3 above)
Comment (olavra) [reply Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs]: Yeah! I chanted it use the multitap option of the input system.

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs (hunk 4 above)
Comment (olavra) [reply Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs:91]: Done!

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs (hunk 5 above)
Comment (olavra) [reply Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs:169]: Done!

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs (hunk 6 above)
Comment (olavra) [reply Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs]: You were right, we have that. Changed.

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs (hunk 7 above)
Comment (olavra) [reply Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs]: I've added the max distance configurable in the CharacterControllerSettings. Regarding the layers I would like th player to be able to click anywhere in the scene, the floor and even the avatars, I dont know the right combination of layers here... :S

File: Explorer/Assets/DCL/Landscape/Assets/GPUI/Grass/StylizedGrass_Indirect.mat (hunk 8 above)
Comment (olavra) [reply Explorer/Assets/DCL/Landscape/Assets/GPUI/Grass/StylizedGrass_Indirect.mat]: Yeah i tried my best to not tick these mat changes in the commits but I failed. Shall we include those in the gitignore?

**Comment by olavra** [comment]
@claude review

**Review by Ludmilafantaniella — APPROVED** [review]
**QA Result: ✅ Approved**

Tested on Windows and Mac.

**What was tested:**
- Double clicking moves the avatar towards the target point correctly
- Marker appears and disappears after reaching the destination or failing to reach it
- Only one marker visible at a time
- Movement is cancelled when using directional keys (WASD)
- Movement is cancelled when opening menus
- Movement is cancelled when the avatar can't reach the destination due to an obstacle
- Shift to run, Space to jump and glide work correctly during automatic movement
- Multiple rapid double clicks handled correctly
- Tested in Genesis City, personal world and scenes with objects
- Controls menu shows the new double click input option

All PR test steps and edge cases verified. No regressions found.

Image

https://github.com/user-attachments/assets/f17553d0-cd9d-49a9-9251-f0f7ac5b6fb2

https://github.com/user-attachments/assets/2bb2cf86-d358-4f37-806e-e18d1da86411

✅Sanity check performed:
- ✔️ Backpack and wearables in world
- ✔️ Emotes in world and in backpack
- ✔️ Teleport with map/coordinates/Jump In
- ✔️ Chat and multiplayer
- ✔️ Profile card
- ✔️ Camera
- ✔️ Skybox
- ✔️ Settings

File: Explorer/Assets/DCL/AvatarRendering/AvatarShape/Assets/DestinationMarker/DestinationMarkerView.cs [hunk 9]
```diff
@@ -0,0 +1,55 @@
+using UnityEngine;
+
+namespace DCL.AvatarRendering.AvatarShape
```
Comment (sandrade-dcl) [inline Explorer/Assets/DCL/AvatarRendering/AvatarShape/Assets/DestinationMarker/DestinationMarkerView.cs]: In my opinion `DestinationMarkerView.cs` doesn't have nothing to do with avatar rendering, I think we should move this class to the `CharacterMotion` namespace (in fact you have the rest of logic already there).

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+
+        private readonly GameObject markerPrefab;
+
+        // Per-frame temp collection for deferred structural changes (allowed by ECS rules).
+        // World.Remove cannot be called while a ref to any component on the same entity is in scope.
+        private readonly List<Entity> entitiesToCancelNavigation = new ();
+
+        private InputAction navigateToAction;
+        private GameObject targetMarker;
+        private DestinationMarkerView targetMarkerView;
```
Comment (alejandro-jimenez-dcl) [inline Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs]: Nitpicking here, I'd use either Destination or Target, but not both

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+        [None(typeof(DeleteEntityIntention))]
+        private void UpdateTarget(
+            [Data] bool hasDoubleClick,
+            [Data] Vector3 clickTarget,
+            [Data] Vector3 clickNormal,
+            Entity entity,
+            in PlayerComponent _)
+        {
+            if (!hasDoubleClick)
+                return;
```
Comment (alejandro-jimenez-dcl) [inline Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs]: we could avoid passing this param and place the query call inside the `hasDoubleClick` check in the `Update`.

**Review by alejandro-jimenez-dcl — APPROVED** [review]
Approved with a few small requests

**Review by sandrade-dcl — CHANGES_REQUESTED** [review]
Amazing feature, good work!! I only left a comment, please take a look to it.

File: Explorer/Assets/DCL/AvatarRendering/AvatarShape/Assets/DestinationMarker/DestinationMarkerView.cs (hunk 9 above)
Comment (olavra) [reply Explorer/Assets/DCL/AvatarRendering/AvatarShape/Assets/DestinationMarker/DestinationMarkerView.cs]: Done!

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs (hunk 11 above)
Comment (olavra) [reply Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs]: Done!

File: Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs (hunk 10 above)
Comment (olavra) [reply Explorer/Assets/DCL/Character/CharacterMotion/Systems/UpdatePointAndClickInputSystem.cs]: ¬¬

Done

**Comment by olavra** [comment]
@alejandro-jimenez-dcl @sandrade-dcl addressed feedback

**Review by sandrade-dcl — APPROVED** [review]

**Comment by olavra** [comment]
@claude review
